libflux/go/libflux/flux.h

```diff
@@ -22,29 +19,53 @@ struct flux_buffer_t {
 // flux_error_t represents a flux error.
 struct flux_error_t;
 
+// flux_error_str will return a string representation of the error.
+// This will allocate memory for the returned string.
+const char *flux_error_str(struct flux_error_t *);
+
+// flux_free will free a resource.
+void flux_free(void *);
+
+// flux_ast_pkg_t is the AST representation of a flux query as a package.
+struct flux_ast_pkg_t;
+
 // flux_parse will take in a string and return the AST representation
 // of the query.
-struct flux_ast_t *flux_parse(const char *);
-
-struct flux_buffer_t *flux_parse_fb(const char *);
+struct flux_ast_pkg_t *flux_parse(const char *);
 
 // flux_ast_marshal_json will marshal json and fill in the given buffer
 // with the data. If successful, memory will be allocated for the data
 // within the buffer and it is the caller's responsibility to free this
 // data. If an error happens it will be returned. The error must be freed
 // using flux_free if it is non-null.
-struct flux_error_t *flux_ast_marshal_json(struct flux_ast_t *, struct flux_buffer_t *);
+struct flux_error_t *flux_ast_marshal_json(struct flux_ast_pkg_t *, struct flux_buffer_t *);
+
+// flux_ast_marshal_fb will marshal the given AST as a flatbuffer into
+// the given buffer. If successful, memory will be allocated for the data
+// within the buffer and it is the caller's responsibility to free this
+// data. If an error happens it will be returned. The error must be freed
+// using flux_free if it is non-null.
+struct flux_error_t *flux_ast_marshal_fb(struct flux_ast_pkg_t *, struct flux_buffer_t *);
 
-// flux_buffer_free will free the memory that was allocated for a buffer.
-// This should only be called if the caller is the one who owns the data.
-void flux_buffer_free(struct flux_buffer_t *);
+// flux_semantic_pkg_t represents a semantic graph package node, including all of its files
+// and their contents.
+struct flux_semantic_pkg_t;
 
-// flux_error_str will return a string representation of the error.
-// This will allocate memory for the returned string.
-const char *flux_error_str(struct flux_error_t *);
+// flux_analyze analyzes the given AST and will populate the second pointer argument with
+// a pointer to the resulting semantic graph. It is the caller's responsibility to free the
+// resulting semantic graph with a call to flux_free().
+// If analysis fails, the second pointer argument wil be pointed at 0, and an error will be returned.
+// Any non-null error must be freed by calling flux_free.
+// Regardless of whether an error is returned, this function will consume and free its
+// flux_ast_pkg_t* argument, so it should not be reused after calling this function.
+struct flux_error_t *flux_analyze(struct flux_ast_pkg_t *, struct flux_semantic_pkg_t **);
```

Comment:
Why does this method use a double pointer? When using `flux_free` on the `flux_semantic_pkg_t`  do I free the first or second pointer? Would it make the API simpler if this were a single pointer?

Reply:
I don't think this would work if it was a single pointer, unless we were to return it as a return value.  But we are returning a `flux_error_t*` and it's nice to be consistent with the other functions.  C doesn't have multiple return values, so we are left with using a function parameter.  (Or creating a new struct type to return both an error and the address of the graph)

Here's why a single pointer wouldn't work.  We want Rust to allocate some memory for a semantic graph, and return a pointer to it.  If we were to pass a single pointer to this function, and have Rust update it to point at the semantic graph, Go would never receive this address, because pointers are passed by value.  To make Go aware of the new value, we need to pass the address of a pointer to a semantic graph, and have Rust update *that*.

The caller needs to free the pointer-to-the-graph, not the pointer-to-the-pointer to the graph.  The latter item is managed by Go.

I agree this is really confusing, and I'm open to other ideas.   This was pretty common practice on C codebases I've worked on.

Reply:
OK thanks for the explanation, maybe we can add the last bit about what to free as a comment and that will be good for me.